src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,858 @@
+    const auto left_min = left_histogram.bin_minimum(left_idx);
+    const auto right_min = right_histogram.bin_minimum(right_idx);
+
+    if (left_min < right_min) {
+      ++left_idx;
+      continue;
+    }
+
+    if (right_min < left_min) {
+      ++right_idx;
+      continue;
+    }
+
+    DebugAssert(left_histogram.bin_maximum(left_idx) == right_histogram.bin_maximum(right_idx),
+                "Histogram bin boundaries do not match");
+
+    const auto height = std::min(left_histogram.bin_height(left_idx), right_histogram.bin_height(right_idx));
+    const auto distinct_count =
+        std::min(left_histogram.bin_distinct_count(left_idx), right_histogram.bin_distinct_count(right_idx));
+
+    if (height > 0 && distinct_count > 0) {
+      builder.add_bin(left_min, left_histogram.bin_maximum(left_idx), height, distinct_count);
+    }
+
+    ++left_idx;
+    ++right_idx;
+  }
+
+  if (builder.empty()) {
+    return nullptr;
+  }
+
+  return builder.build();
+}
+
+std::shared_ptr<TableStatistics> CardinalityEstimator::estimate_inner_equi_join(
+    const ColumnID left_column_id, const ColumnID right_column_id, const TableStatistics& left_input_table_statistics,
+    const TableStatistics& right_input_table_statistics) {
+  const auto left_data_type = left_input_table_statistics.column_data_type(left_column_id);
+  const auto right_data_type = right_input_table_statistics.column_data_type(right_column_id);
+
+  // TODO(anybody) - Implement join estimation for differing column data types
+  //               - Implement join estimation for String columns
+  if (left_data_type != right_data_type || left_data_type == DataType::String) {
+    return estimate_cross_join(left_input_table_statistics, right_input_table_statistics);
+  }
+
+  std::shared_ptr<TableStatistics> output_table_statistics;
+
+  resolve_data_type(left_data_type, [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+
+    const auto left_input_column_statistics = std::dynamic_pointer_cast<ColumnStatistics<ColumnDataType>>(
+        left_input_table_statistics.column_statistics[left_column_id]);
+    const auto right_input_column_statistics = std::dynamic_pointer_cast<ColumnStatistics<ColumnDataType>>(
+        right_input_table_statistics.column_statistics[right_column_id]);
+
+    auto cardinality = Cardinality{0};
+    auto join_column_histogram = std::shared_ptr<AbstractHistogram<ColumnDataType>>{};
+
+    auto left_histogram = left_input_column_statistics->histogram;
+    auto right_histogram = right_input_column_statistics->histogram;
+
+    if (left_histogram && right_histogram) {
+      join_column_histogram = estimate_inner_equi_join_with_histograms(*left_histogram, *right_histogram);
+      cardinality = join_column_histogram->total_count();
+    } else {
+      // TODO(anybody) If there aren't histograms on both sides, use some other algorithm/statistics to estimate the
```

Comment:
What did we do before? Why can't we continue to do that?

Reply:
Before we probably just crashed/failed if there were no statistics available for an operation? There is nothing to "continue" doing, hence.
Atm there will be histograms for all tables and all operations so this TODO is mainly to highlight the place where other cardinality-estimating statistics can be hooked in.